sql: scan the system.role_members table using a single scan

rafiss opened the pull request:
relates to support case https://github.com/cockroachlabs/support/issues/1463
fixes https://github.com/cockroachdb/cockroach/issues/77452

### sql: use singleflight for role membership lookups

This helps prevent a thundering herd when this cache gets invalidated.

### sql: scan the system.role_members table using a single scan


Release note (sql change): Added a
`sql.auth.resolve_membership_single_scan.enabled` cluster setting, which
changes the query for an internal role membership cache. Previously the
code would recursively look up each role in the membership hierarchy,
leading to multiple queries. With the setting on, it uses a single
query. The setting is true by default.

### sql: fix information_schema role introspection

Release note (bug fix): The information_schema tables
administrable_role_authorizations and applicable_roles were incorrectly
always returning the current user for the grantee column. Now, the
column will contain the correct role that was granted the parent role
given in the role_name column.

Release justification: high priority bug fix

otan committed:
sql: scan the system.role_members table using a single scan

Release justification: high priority bug fix gated by a cluster setting

Release note (sql change): Added a
`sql.auth.resolve_membership_single_scan.enabled` cluster setting, which
changes the query for an internal role membership cache. Previously the
code would recursively look up each role in the membership hierarchy,
leading to multiple queries. With the setting on, it uses a single
query. The setting is true by default.

rafiss committed:
sql: use singleflight for role membership lookups

This helps prevent a thundering herd when this cache gets invalidated.

Release justification: high priority bug fix
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/77359)
<!-- Reviewable:end -->

otan commented on pkg/sql/user_test.go:
```diff
@@ -48,190 +49,199 @@ func TestGetUserTimeout(t *testing.T) {
-			if _, err := dbSQL.Exec(ctx, "SELECT 1"); err != nil {
-				t.Fatal(err)
+			if singleScan {
+				_, err := db.Exec(`SET CLUSTER SETTING sql.auth.resolve_member_single_scan.enabled = true`)
```
Maybe take out the if statement

rafiss replied:
done

otan reviewed (commented):
Quick query: are we planning to backport this?

rafiss commented:
yeah, i was planning to backport, but change the cluster setting to default to false in 21.2

rafiss committed:
sql: fix information_schema role introspection

Release justification: low risk bug fix

Release note (bug fix): The information_schema tables
administrable_role_authorizations and applicable_roles were incorrectly
always returning the current user for the grantee column. Now, the
column will contain the correct role that was granted the parent role
given in the role_name column.

ajwerner reviewed (commented):
> Reviewed 1 of 4 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77359)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @otan, @rafiss, and @RichardJCai)

___
*[pkg/sql/authorization.go, line 497 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/77359#-MxXgAu42pKE5Q-fc3KW:-MxXgAu42pKE5Q-fc3KX:b-frzccn) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a1e611b69650706917541c44492e77fdb752cc0/pkg/sql/authorization.go#L497)):*
> ```Go
>         INNER JOIN cte_members AS child ON parent.member = child.role
>     )
> SELECT * FROM cte_members`
> ```

is the plan here actually good? Is this really a good idea? On some level this seems great, but on another, we've already committed to keeping the whole cache in RAM. :shrug: 



_Code quote:_
```Go

  cte_members (member, role, "isAdmin")
    AS (
      SELECT member, role, "isAdmin"
      FROM system.role_members
      WHERE member = $1
      UNION ALL
        SELECT parent.member, parent.role, parent."isAdmin"
        FROM system.role_members AS parent
        INNER JOIN cte_members AS child ON parent.member = child.role
    )
SELECT * FROM cte_members`
```
___
*[pkg/sql/authorization.go, line 498 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/77359#-MxXgUo04sw6oIxnYCJR:-MxXgUo04sw6oIxnYCJS:b4h0yv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a1e611b69650706917541c44492e77fdb752cc0/pkg/sql/authorization.go#L498)):*
> ```Go
>     )
> SELECT * FROM cte_members`
> 		it, err := ie.QueryIterator(
> ```

does this need a defer `it.Close`? I can see the argument why it doesn't, namely there's not other early error return, but I'd rather you put it because of panics and future code changes. 
___
*[pkg/sql/information_schema.go, line 2675 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77359#-MxXgxQdE0lvFgg96V88:-MxXgxQdE0lvFgg96V89:b-bbacik) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8400cab8b0620cd8f7c70c75a47427d6cc2dca/pkg/sql/information_schema.go#L2675)):*
> ```Go
> }
> 
> func forEachRoleHierarchyOfUser(
> ```

it feels like we should factor this out 


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77359)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @otan, and @RichardJCai)

___
*[pkg/sql/authorization.go, line 497 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/77359#-MxXgAu42pKE5Q-fc3KW:-Mx_L4LR82Pj-4TX8L3W:b-a2rerh) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a1e611b69650706917541c44492e77fdb752cc0/pkg/sql/authorization.go#L497)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

is the plan here actually good? Is this really a good idea? On some level this seems great, but on another, we've already committed to keeping the whole cache in RAM. :shrug: 
</blockquote></details>

I'm experimenting a bit with sample data to see what works best
___
*[pkg/sql/authorization.go, line 498 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/77359#-MxXgUo04sw6oIxnYCJR:-Mx_PcQJEeZH3tLrQ94v:bg796yz) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a1e611b69650706917541c44492e77fdb752cc0/pkg/sql/authorization.go#L498)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

does this need a defer `it.Close`? I can see the argument why it doesn't, namely there's not other early error return, but I'd rather you put it because of panics and future code changes. 
</blockquote></details>

makes sense, adding.
___
*[pkg/sql/information_schema.go, line 2675 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77359#-MxXgxQdE0lvFgg96V88:-Mx_Peaw-f2Up0oFmiIs:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8400cab8b0620cd8f7c70c75a47427d6cc2dca/pkg/sql/information_schema.go#L2675)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

it feels like we should factor this out 
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

otan commented on pkg/sql/authorization.go:
```diff
@@ -464,14 +465,58 @@ func MemberOfWithAdminOption(
 	ret := map[security.SQLUsername]bool{}
+	if singleQuery {
+		const lookupRolesStmt = `
+WITH RECURSIVE
```
this is nice! i wonder if we should test false cases more though by, e.g. running all of the `role` logic tests with true and false (maybe by defining a new logic test setup like `5node`)

otan reviewed (approved):
looking at first and third commits only (second is a separate PR i believe)
i would appreciate that extra test mentioned

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77359)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @otan, @rafiss, and @RichardJCai)

___
*[pkg/sql/authorization.go, line 497 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/77359#-MxXgAu42pKE5Q-fc3KW:-MxaR9A24WjRIygRPGpA:b-1cn08) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a1e611b69650706917541c44492e77fdb752cc0/pkg/sql/authorization.go#L497)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

I'm experimenting a bit with sample data to see what works best
</blockquote></details>

I did a few benchmarks. I created 150 users and randomly added 500 memberships between them. That's not super realistic, since it leads to a role hierarchy with many levels, but it showed that the performance of recursive CTEs can quickly get out of control.

Also, it helped me see that my first attempt at a recursive CTE has a bug, since `UNION ALL` would lead to duplicate rows being included in the output. Adding a `DISTINCT` in the outer SELECT query fixes it, but that makes performance even worse. Using `UNION` (without all) fixes the duplicates bug and performs better than `UNION ALL`, but it's only available in v22.1.

```
BenchmarkLotsOfRoles/full_scan         	     338	   3305653 ns/op

BenchmarkLotsOfRoles/recursive_cte_union         	     157	   7131301 ns/op

BenchmarkLotsOfRoles/recursive_cte_union_all    did not complete
```


<!-- Sent from Reviewable.io -->

ajwerner commented:
@rafiss can you try out this query on your data:

```sql
WITH RECURSIVE cte_members (role, path, admin, cycle) AS (
                         SELECT role, ARRAY[$1], "isAdmin", false
                           FROM system.role_members
                          WHERE member = $1
                         UNION ALL SELECT parent.role,
                                          child.path || parent.member,
                                          parent."isAdmin",
                                          parent.member = ANY (child.path)
                                     FROM system.role_members AS parent
                               INNER JOIN cte_members AS child ON parent.member
                                                                  = child.role
                                                              AND NOT cycle
                     )
        SELECT $1, role, bool_or(admin) AS "isAdmin" -- note we don't need to return the member back
          FROM cte_members
      GROUP BY role;
```

rafiss commented:
That query also is taking a long time (it's been running more than 1 minute).

otan commented:
should we just use a single SQL scan then?

rafiss commented:
> should we just use a single SQL scan then?

yep, done

otan commented on pkg/sql/authorization.go:
remind me - should we wrap this error?

otan commented on pkg/sql/logictest/testdata/logic_test/role:
```diff
@@ -181,18 +181,33 @@ user root
 statement ok
```
can we run this logic test file with cluster setting on and off?
add something to https://github.com/cockroachdb/cockroach/blob/c6a5a395cbecbbd5ba77c2c14e3bf68eb59eeb19/pkg/sql/logictest/logic.go#L679-L684 and add `LogicTest: local .... local-cache-single-read-off` or something?

rafiss replied:
added a metamorphic constant

otan reviewed (approved):
(lgtm otherwise but i'd love that test)

rafiss commented:
I changed the default value of `sql.auth.resolve_membership_single_scan.enabled` to be a metamorphic constant, so test builds have a chance of defaulting to false.

rafiss commented:
bors r=otan
